added the Simpoint Class and simpoints exit event support on set_se_workload and Simulator

studyztp opened the pull request:
Added a new Simpoint class under the src/python/gem5/components/processor/simpoint.py
It can take in either the simpoint and weight file paths or the simpoint and weight list.

Added the checkpoint generator and dump_reset generator.

The set_se_workload can take in Simpoint object and use the get function to pass the simpoint_start_insts to the core.

Added a schedule_simpoints function for scheduling simpoints in fs mode and schedule simpoint end if needed in Simulator.
Also added a schedule restore_simpoint_stop function in Simulator to schedule a stop when doing restore simpoint checkpoints.

Commit:
arch-sparc: Implement RegClass based register flattening.

Change-Id: Ib1d5b7ac7ff5aa6f35099fd9fd4530951c3efa19
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51230
Maintainer: Gabe Black <gabe.black@gmail.com>
Reviewed-by: Matthew Poremba <matthew.poremba@amd.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
arch,cpu: Replace calls to (get|set)RegFlat.

Make these use RegIds which are based on already flattened RegClass-es.

Change-Id: I50f50614830c7010c18a8ebb95aba8decc078ac0
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51231
Maintainer: Gabe Black <gabe.black@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
cpu: Eliminate the (get|set)RegFlat methods.

These can now be performed with the (reg|set)Reg methods by using an
already flattened RegId.

Change-Id: Ie02cda224d96644061227eada100675d38797e57
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51232
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Maintainer: Gabe Black <gabe.black@gmail.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
arch-arm,cpu: Remove all uses of flattenRegId.

RegIds can now be flattened directly.

Change-Id: I2a603c12bbc586720082363996f303cd3b43ac9c
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51233
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
fastmodel,cpu: Remove the flattenRegId ThreadContext method.

This is no longer used or necessary.

Change-Id: Ide8dd74b5d39b245d3d71979dd84c7fee60d566e
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51234
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
arch: Remove unused register flattening methods.

Most flattening methods in the ISA classes aren't used. The one
exception is the flattenMiscIndex function in ARM.

Change-Id: I26088692fe3f56914009afb0203f59da7cf6023a
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51235
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
arch: Make the ISA::clear() method virtual.

This method is not at all on the critical path, and the adding virtual
method overhead to it will have essentially no effect on over all
performance.

Change-Id: I583bbe30f5ed923a0c771f36c3f07e5979c28476
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51236
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
arch: Make the (read|set)MiscReg methods virtual.

They will then be accessible through the base class. This adds some
small hypothetical amount of overhead to each call, but a quick test
booting linux on x86 does not show any significant slowdown.

Change-Id: If706ddf173d4e24d85468fb118b45099acf3c05b
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51237
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
cpu: Store the ISA class using BaseISA and not TheISA::ISA.

All generic functionality of the ISA class can now be accessed using
virtual methods, and so we don't need to keep the ISA specific version
of the class around any more.

Change-Id: I9f9a3de2dc68983276ef7efc008a18960412d509
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51238
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
cpu: Stop including arch/isa.hh.

This header file is no longer needed.

Also fix some places where the isa header file was being transitively
included.

Change-Id: Ib9a9d7db0c9808b29d7614bbd68e2052ea345e9f
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51239
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
scons: Get rid of the isa.hh switching header file.

Change-Id: I1aebf05f6d0f2cde3841d6e6d3361c9733538b08
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/51240
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
misc: Stop excluding code when building the NULL ISA.

The BaseCPU needs a little extra hacking because it tries to create
default objects based on what the ISA is. If the ISA isn't recognized,
then the types will be set to None, and some extra checks have been
added as the type is set up.

Change-Id: Ia3cae313e1a96a953d2316d9192f41a8fd28c141
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62195
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

Commit:
misc: Remove lingering uses of TheISA::.

Change-Id: Ie55e0d79867fbc8f75a993fb456a58c84de5def4
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62196
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>

Commit:
misc: Stop including config/the_isa.hh.

It is no longer necessary anywhere in gem5.

Change-Id: Iac999acf8c59ee7387214057bebb617acd01617c
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62197
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Gabe Black <gabe.black@gmail.com>
Maintainer: Gabe Black <gabe.black@gmail.com>

Commit:
scons: Remove support for generating config/the_isa.hh.

Change-Id: I06e4fa0193009fe1a84d8b24af4fe3f09057e674
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62198
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Maintainer: Giacomo Travaglini <giacomo.travaglini@arm.com>

alexdutu committed:
gpu-compute: Adding support for LDS atomics

This changeset is adding support for LDS atomics
and implementing DS_OR_B32 instruction.

Change-Id: I84c5cf6ce0e9494726dc7299f360551cd2a485f5
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61791
Reviewed-by: Matt Sinclair <mattdsinclair@gmail.com>
Maintainer: Matt Sinclair <mattdsinclair@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

aakahlow committed:
stdlib: Add a component for HBM2 stack

This change adds a component for HBM2 stack in the gem5 stdlib.
For HBM2 stack, the atom size is used to interleave across pseudo
channels in a single physical channel or HBMCtrl and the bits
beyond that will be used to interleave across channels/controllers.

Change-Id: I95a279504981a5c000f38c9a6ad0e03484eb258e
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61489
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
stdlib: Fix SimpleSwitchableProcessor to allow Minor type

Without setting the correct memory mode the SimpleSwitchableProcessor,
the Minor CPU could not be used as a valid core. This patch corrects
this issue by setting the memory mode to TIMING for Minor CPU cores.

Due to the increasingly complex if-else to determine the memory mode, a
function has been added to CPUTypes to determine what MemMode is
required for each CPUType.

Change-Id: I9384b4a9c0673af34cca04917d763ca45d0ea434
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61535
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
arch-mips,cpu-minor: Add MinorCPU to MIPS ISA

While it may not be well supported, it's better to incorporate the
MinorCPU into the MIPS ISA gem5 binary than leave it out.

Change-Id: If44aa0531f287f4c3d8789c54025c7bb5259586a
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61536
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
arch-power,cpu-minor: Add MinorCPU to POWER ISA

While it may not be well supported, it's better to incorporate the
MinorCPU into the POWER ISA gem5 binary than leave it out.

Change-Id: I408b0329ff2633e70995e9751ac4c6715dd38cb7
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61537
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
arch-mips,cpu-minor: Add MinorCPU to SPARC ISA

While it may not be well supported, it's better to incorporate the
MinorCPU into the SPARC ISA gem5 binary than leave it out.

Change-Id: Iff8016b2a9857888cf66d1d2060581e979111b9f
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61538
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
arch-mips,cpu-minor: Add MinorCPU to X86 ISA

While it may not be well supported, it's better to incorporate the
MinorCPU into the X86 ISA gem5 binary than leave it out.

Change-Id: I98d015c23e347276f6b943291877dc15026eb0da
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61539
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

aakahlow committed:
mem: Fix the respondEvent check in DRAM interface

This change fixes the respondEvent schedule check
in DRAM interface to make sure that the correct respondEvent
is checked for a given pseudo channel.

Change-Id: Ie5edf48db9f6cf2e7ee4cafe7b774441464d77a4
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61629
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

studyztp committed:
tesging

studyztp committed:
added simpoint support for set_se_workload and simulator

BobbyRBruce committed:
stdlib: Fix incorrect return type in cpu_types and isa

`get_cpu_types_str_set()` and `get_isas_str_set()` return `Set[str]` not
`Set[CPUTypes]`/`Set[ISA]`.

Change-Id: I703ce4c19e77eb6a3931cabb759f25d28aabb412
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/61773
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

studyztp committed:
stdlib: take and create exit event for simpoint

Change-Id: I26b447801e22a51a182483faf878b580e53f5799

studyztp committed:
stdlib: kernal set workload takes simpoints

Change-Id: Id8e2c466571a63c7671b7eafab8080834cc4b44e

gzaets committed:
python: adding BBV generator config

Change-Id: I2b711fc5b065c6995b7a3dd9246f4b61e4711e86

studyztp committed:
stdlib: simulator() includes simpoint schedule functions

Change-Id: Ib5b99f5edff6b930400e2921f5ebb53934154197

BobbyRBruce committed:
tests: Fix incorrect doc-string in test_hello_se.py

Change-Id: Ic21e59cfc8fba92f93661d8b33803bd4c4c6fc58
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62011
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
tests: Add "tests/gem5/se_mode" directory for SE tests

At present only the "test_hello_se.py" tests are run, but this directory
will serve as a directory for other SE tests going forward.

Change-Id: I814d1e673df0195960d3af69277db8d397f299e5
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62012
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
stdlib,tests: Fix stdlib SE mode for multicore setups

The `set_se_binary_workload` function was only setting up the binary to
work on one (the first) processor core. This caused an exception to be
thrown when trying to run an SE mode binary on a multicore system.

Tests have been added to ensure this works as intended.

Note: While this implementation fixes the bugs, it is limited. Future
work is needed to allow for multiprogram workloads.

Change-Id: I33dbaf5015705c299215dc83e8449b16df301cd4
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62014
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>

studyztp committed:
stdlib: added simpoint class and simpoint exit event support

Change-Id: I1407882d4d8c120ea5ee2bad9f209f81bd50e8e3

powerjg commented:
Excited to check this out! Hopefully, I'll get a chance this afternoon :)

studyztp commented:
> 

Thank you!! No rush. :)

studyztp committed:
Update kernel_disk_workload.py

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+
```
Needs a license

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+
+
+from m5.util import *
```
It's better to avoid `import *` if possible.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+        :param weight_file_path: the path to the weight result file generated
+        by Simpoint3.2 or gem5
+
+        :param simpoint_list: a list of simpoint starting instructions (sorted)
```
Sorted how? What us the sorting?

studyztp replied:
The simpoint_list is a list of the simpoint interval numbers (ex. [3, 5, 15]). Sorry about the names, I'm bad at naming. They should be sorted in ascending order. This might be good for cases that need to schedule starting instructions one by one during the simulation. But now I think about it, you're right, they don't need to be sorted.

powerjg replied:
I see what you're saying now. Sorted based on the instruction number, not based on the weights or anything else.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+
+        """
+
+        if simpoint_interval is None:
+            fatal("not passing in interval length")
```
If it's required, then it shouldn't have a default None value :)

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+        if simpoint_file_path is None or weight_file_path is None:
+            if simpoint_list is None or weight_list is None:
+                # raise error
+                fatal("not passing in enough information")
```
A more specific error would be good.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+        simpoint_path:Path,
+        weight_path:Path,
+        )-> Tuple[List[int],List[int]]:
+        simpoint_file = open(simpoint_path)
+        weight_file = open(weight_path)
```
I would suggest using `with open(filename) as simpoint_file:`

https://realpython.com/python-with-statement/#:~:text=The%20Python%20with%20statement%20creates,finally%20statement.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,99 @@
+        # use simpoint to sort
+        simpoint_start_insts = []
+        weight_list = []
+        for obj in simpoint:
```
```suggestion
        for start,weight in simpoint:
```

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -95,3 +98,29 @@ def default_workend_generator():
+    while True:
+        # only print for testing purpose
+        print("got to the simpoint_save_checkpoint_generator\n")
+        m5.checkpoint(joinpath(checkpoint_dir,f"cpt.{str(m5.curTick())}"))
```
We should think more about how to do this. I don't love that we're using the `m5` interface here. @BobbyRBruce, I think that we need to have a way to get to the simulator instance from these generators.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +190,18 @@ def unique_exit_event():
+        self._board.get_processor().get_cores()[0].\
+                        set_simpoint(simpoint_end_inst,simpoint_start_insts)
+
+    def schedule_simpoint_restore_stop(self, simpoint_interval: int):
```
What is this function used for? Could you add pydoc for this and the function above?

powerjg reviewed (commented):
Generally, this looks great to me :)

studyztp committed:
Update src/cpu/base.cc

Co-authored-by: Jason Lowe-Power <jason@lowepower.com>

studyztp commented on src/cpu/base.cc:
```diff
@@ -727,6 +727,28 @@ BaseCPU::traceFunctionsInternal(Addr pc)
+        }
+    }
+    else {
+        scheduleInstStop(0, end_point-1, cause2);
+        // problem might occur when starting point = [0,1], because for inst 0,
+        // the event are scheduled at inst 1, and when inst 1 schedule an end
+        // point, the end point event will get scheduled at 1*interval length
```
When I was doing testing, I encountered a conflict problem with non-KVM CPU when passing in simpoint_start_instruction = [0, 100000000] and schedule an end point when it received the SIMPOINT_BEGIN exit event. The end point event for simpoint 0 was raised after the SIMPOINT_BEGIN for simpoint 100000000 at instruction 100000000. I found that the SIMPOINT_BEGIN for simpoint 0 was scheduled at instruction 1, so when I scheduled an end point event for simpoint 0, it was schedule at instruction 1+end_point-1. In that case, there are two events scheduled at instruction 100000000, and it caused a conflict.

studyztp commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -110,6 +110,26 @@ def get_mmu(self) -> BaseMMU:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_simpoint(
```
The set_simpoint is for scheduling events after the simulation starts, and init_simpoint is for scheduling event before the simulation start. The use case for init_simpoint is the set_se_workload and restoring checkpoints. For example, when restoring a checkpoint, there isn't a beginning exit event for us to schedule the simpoint, so we need to schedule it before the simulation starts. Sorry about not having the comments and documentations, I'll add them and push them.

powerjg replied:
This makes sense on the C++ side, but I wonder if it's absolutely needed on the python side. Could you detect that it's "init" and do the right thing.

studyztp replied:
Yes, I can add a boolean parameter to detect it. And I've been thinking should it have a boolean parameter to allow user to choose between scheduling SIMPOINT_BEGIN or SIMPOINT_END?

powerjg replied:
The SIMPOINT_END is just SIMPOINT_BEGIN + interval_length, right?

I guess I'm a bit confused by that as well.

studyztp replied:
I think of SIMPOINT_END as an exit event that's simpoints related but distinguish from the SIMPOINT_BEGIN. I can only think of use cases like doing switching CPUs in FS mode. Maybe I should get rid of it and change SIMPOINT_BEGIN to SIMPOINT_FOUND, so there is only one exit event for simpoints to make it less confusing.

powerjg replied:
I definitely like SIMPOINT_BEGIN. However, maybe the "SIMPOINT_END" should just be "number of instructions reached" or whatever that exit event is...

studyztp commented on src/cpu/base.hh:
```diff
@@ -178,6 +178,10 @@ class BaseCPU : public ClockedObject
     virtual Port &getDataPort() = 0;
 
+    void scheduleSimpoint(
+      Counter end_point,
```
The end_point is an instruction number. When user pass in an end_point, it will schedule a SIMPOINT_END exit event at the end_point+current local instructions.

studyztp committed:
stdlib: support MAX_INSTS and schedule exit events

BaseCPU.py
Added scheduleOneMaxInsts

base.cc&base.hh
Moved original max instruction count event to scheduleOneMaxInsts
Removed support for SIMPOINT_END

se_workload.py
Added comments about the simpoint parameter
Removed some old simpoint comments

abstract_core.py&simple_core.py
Added set_one_max_insts function to schedule for MAX_INSTS
Added new parameters init for both scheduling functions to choose
when to schedule the events
Added pydoc comments
Removed init_simpoint function and end_point parameter

simpoint.py
Added license
Changed to using with statement to open files
Modified comments and variable names

exit_event.py
Added MAX_INSTS

simulation.py
Added schedule_max_insts function
Added pydoc comments for new scheduling functions

Change-Id: Ie655c95fd8bff3669c7ea06c14f69c0f758451aa

studyztp committed:
Update base.cc

powerjg commented on src/cpu/base.hh:
```diff
@@ -443,7 +443,28 @@ class BaseCPU : public ClockedObject
+     * @param inst_starts A vector of number of instructions to start simpoints
+     */
+
+    void scheduleSimpoint(std::vector<Counter> inst_starts);
```
```suggestion
    void scheduleSimpointsInstStop(std::vector<Counter> inst_starts);
```

powerjg commented on src/cpu/base.hh:
```diff
@@ -443,7 +443,28 @@ class BaseCPU : public ClockedObject
+     *
+     * @param max_insts Number of instructions into the future.
+     */
+    void scheduleOneMaxInsts(Counter max_insts);
```
```suggestion
    void scheduleInstStopAnyThread(Counter max_insts);
```

powerjg commented on src/cpu/base.cc:
```diff
@@ -727,6 +723,21 @@ BaseCPU::traceFunctionsInternal(Addr pc)
+BaseCPU::scheduleOneMaxInsts(Counter max_insts)
+{
+    std::string cause = "a thread reached the max instruction count";
+    for (ThreadID tid = 0; tid < numThreads; ++tid)
+        scheduleInstStop(tid, max_insts, cause);
```
```suggestion
    for (ThreadID tid = 0; tid < numThreads; ++tid)
        scheduleInstStop(tid, max_insts, cause);
    }
```

powerjg commented on configs/example/BBV/bbvprobe.py:
```diff
@@ -0,0 +1,100 @@
+    # Any resource specified in this file will be automatically retrieved.
+    # At the time of writing, this file is a WIP and does not contain all
+    # resources. Jira ticket: https://gem5.atlassian.net/browse/GEM5-1096
+    Resource("arm-hello64-static")
```
It says x86 above :)

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,117 @@
+# Copyright (c) 2021 The Regents of the University of California
```
I'm not sure this file makes sense in the "processors" package. Maybe in `util`? @BobbyRBruce, any opinion?

BobbyRBruce replied:
Yeah. I don't think this file has a logical home, but i think `util` makes the most sense here.


Also, the copyright should be 2022, not 2021.

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -54,6 +60,10 @@ def set_se_binary_workload(
         items. True by default.
+        :param stdin_file: The input file for the binary
+        :param arguments: The input arguments for the binary
+        :param simpoint: The SimPoint object that contains the list of simpoint
```
```suggestion
        :param simpoint: The SimPoint object that contains the list of simpoints
```

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -64,11 +74,21 @@ def set_se_binary_workload(
+
         self.get_processor().get_cores()[0].set_workload(process)
 
+        if simpoint is not None:
```
We should probably add a warning that SimPoints only works with one core.

powerjg commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -110,6 +110,32 @@ def get_mmu(self) -> BaseMMU:
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_one_max_insts(
```
See name suggestion in `base.hh`

powerjg commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -110,6 +110,32 @@ def get_mmu(self) -> BaseMMU:
+    @abstractmethod
+    def set_simpoint(
+        self,
+        inst_starts: List[int] = [],
+        init: bool = True
```
Define these parameters :). Especially the `init` parameter.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,117 @@
+        # with it in the future
+    ) -> None:
+        """
+        :param simpoint_interval: the length of each simpoint interval
+        :param simpoint_file_path: the path to the simpoint result file
+        generated by Simpoint3.2 or gem5
+        :param weight_file_path: the path to the weight result file generated
+        by Simpoint3.2 or gem5
+
+        :param simpoint_list: a list of simpoint starting instructions
+        :param weight_list: a list of weights
```
Make it clear that you need either the path or the list in this help documentation.

powerjg commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,117 @@
+        return self._weight_list
+
+    def get_simpoint_interval(self)->int:
+        return self._simpoint_interval
```
empty newlines at the end of the file are nice

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -28,6 +28,9 @@
 from ..components.processors.abstract_processor import AbstractProcessor
 from ..components.processors.switchable_processor import SwitchableProcessor
 from m5.util import warn
+from os.path import join as joinpath
+from pathlib import Path
```
Do we need both of these?

If using pathlib, I think you can use this instead: https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -28,6 +28,9 @@
 from m5.util import warn
+from os.path import join as joinpath
+from pathlib import Path
+import os
```
Is this used?

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +190,38 @@ def unique_exit_event():
+        :param schedule_at_init: if it is True, schedule the events in the init
+        stage of the core, else, schedule the events during the simulation
+        """
+        self._board.get_processor().get_cores()[0].\
```
Warn if more than 1 core?

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +190,38 @@ def unique_exit_event():
+        stage of the core, else, schedule the event during the simulation
+        """
+        self._board.get_processor().get_cores()[0].\
+                        set_one_max_insts(inst,schedule_at_init)
```
```suggestion
                        set_one_max_insts(inst, schedule_at_init)
```

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +190,38 @@ def unique_exit_event():
+        self._board.get_processor().get_cores()[0].\
+                        set_simpoint(simpoint_start_insts,schedule_at_init)
+
+    def schedule_max_insts(
```
For this function, we definitely don't want to assume just 1 core. I'm not certain I know *what* we want to do, though.

I think maybe this should be a board function / processor function.

studyztp replied:
Maybe it should be a processor function because it's more related to a processor than to a board..?

powerjg replied:
I'll drop this comment. I'm trying to look for too much "perfection" and I'm not really sure there's a better way.

powerjg reviewed (commented):
Can you add an example runscript that uses simpoints? I think this looks good, but I'd like to see the external user interface.

studyztp commented:
> Can you add an example runscript that uses simpoints? I think this looks good, but I'd like to see the external user interface.

Yes, I'll add an example runscript. I'll push it later with the fixed files and ping you and Bobby. Thank you for reviewing :)

studyztp committed:
stdlib: modified SimPoints supports

fs_checkpoint.py & fs_restore.py
two example scripts for using SimPoints related functions
00000000000000000000000000000000000000000000000000000000000000000000000000
BaseCPU.py, base.cc & base.hh
Changed the name of the schedule function for MAX_INSTS to
scheduleInstStopAnyThread

se_binary_workload.py
Added warning when using SimPoints functions with more than one core

abstract_core.py & simple_core.py
Added pydoc lines to explain the parameters for the SimPoints and
MAX_INSTS schedule functions
Changed the name for the MAX_INTS schedule function to
set_inst_stop_any_thread

simpoint.py
Added an empty line at the end of the file
Added usage note to ask for either paths or lists for both SimPoints and
weights

exit_event_generators.py
Removed unused libraries
Changed to using pathlib in the SimPoints checkpoint generator

simulator.py
Added warning when using SimPoints functions with more than one core

Removed the bbvprobe.py because it was a testing file

Change-Id: I1f353707cbeeb7f47d3df331a7e01df1807f7093

studyztp committed:
Update fs_restore.py

studyztp committed:
configs: added comments to the configs scripts

Change-Id: Iba78f2097eaecc96d9fb08041e8324442a4fddd8

studyztp committed:
Rename fs_checkpoint.py to simpoints-fs-checkpoint.py

studyztp committed:
Rename fs_restore.py to simpoints-fs-restore.py

yuxiaojia committed:
mem-ruby: Fix replacement policy updates in MI_example

The current MI_example protocol's L1 caches updates the MRU information twice per request on misses -- once when the request reaches Ruby and once when the miss is returned from another level of the memory hierarchy.

Although this approach does not cause any correctness bugs for replacement policies like LRU since this request is the LRU in both cases, it does not work correctly for other policies like SecondChance and LFU, where updating the information twice (for misses) causes them to devolve to LRU.

Note that this was not directly a problem with Ruby previously, because it only supported LRU-based policies that were unaffected by this.  However, with the integration of 20879 Ruby now uses the same replacement policies as Classic (which has additional, non-LRU based replacement policies).

This patch resolves this problem by not updating the MRU information a second time for the misses. It has been tested and validated with the replacement policy tests in 20880.

Change-Id: I82a57abf2a16d70820413ba8118378f2e91fd7fb
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62232
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Matt Sinclair <mattdsinclair@gmail.com>
Reviewed-by: Matt Sinclair <mattdsinclair@gmail.com>

studyztp committed:
stdlib, configs: changed examples from FS to SE

Fixed some grammar issues in comments and pydocs
Changed the example configs scripts to SE mode

Change-Id: I9e65a16c2415a2ae2d8dfad326010d1fe1bc14de

studyztp committed:
Rename simpoints-fs-checkpoint.py to simpoints-se-checkpoint.py

studyztp committed:
Rename simpoints-fs-restore.py to simpoints-se-restore.py

studyztp committed:
stdlib: fixed some style issues with simpoint.py

Change-Id: Ib25b369cf01384939db49a9a0b828e9f749dafe6

studyztp committed:
fixed a mistake line in the simpoint-se-restore.py

studyztp committed:
configs: removed commands

Change-Id: Id0a219c9fdf8c3c46f180aac2e8c4f40eb2182ec

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,70 @@
+
+requires(isa_required=ISA.X86)
+
+cache_hierarchy = PrivateL1PrivateL2CacheHierarchy(
```
I suggest using `NoCache` when taking checkpoints for SimPoints (at least in our examples). When restoring from a checkpoint, there's no cache state saved, so you can completely change the cache hierarchy. By using `NoCache` to take the checkpoints, we'll get slightly better performance when running in atomic mode, and it won't put any restrictions on what people can do *with* the checkpoints. 

As an aside, when we finalize these scripts as examples, we should write things like the above in comments so everyone can learn.

studyztp replied:
I'll change to NoCache and add the comments. It's a very useful information. Thank you!

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,70 @@
+    l2_size="256kB",
+)
+
+memory = DualChannelDDR4_2400(size = "3GB")
```
Similarly, we can use `SimpleMemory` to (possibly) improve performance.. Though, TBH, I'm not sure this makes any difference.

studyztp replied:
When I tried to use SimpleMemory, there is an error caused by it not being inherited from the AbstractMemorySystem. It doesn't have the 'get_size' function the stdlib board needs during the its init stage. So I used the SingleChannelDDR3_1600 when taking checkpoints, and changed to DualChannelDDR4_2400 when restoring from a checkpoint. Does it sounds correct?

studyztp replied:
During the testing, I also encountered an error when I had two different sizes setup for the memory when taking checkpoints and restoring from one. Does it mean the memory size has to be the same?

powerjg replied:
Yes and yes 🙂

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,70 @@
+processor = SimpleProcessor(
+    cpu_type=CPUTypes.ATOMIC,
+    isa=ISA.X86,
+    num_cores=1,
```
Perfect. I would note that SimPoints requires 1 core.

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,70 @@
+dir = Path("se_checkpoint_folder/")
+dir.mkdir(exist_ok=True)
+
+simulator = Simulator(
+    board=board,
+    on_exit_event={
+        # using the SimPoints event generator in the standard library to take
+        # checkpoints
+        ExitEvent.SIMPOINT_BEGIN: simpoint_save_checkpoint_generator(dir)
+    }
+)
```
I LOVE THIS! I'm sooo happy seeing this code!!

studyztp replied:
XD

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,70 @@
+from gem5.components.processors.simpoint import SimPoint
+from pathlib import Path
+
+cache_hierarchy = PrivateL1PrivateL2CacheHierarchy(
```
*Here* is a good place to use the more complex hierarchy. This is great. Make a note in a comment that you can change the hierarchy :)

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,70 @@
+
+requires(isa_required=ISA.X86)
+
+memory = DualChannelDDR4_2400(size = "3GB")
```
Same as above. This is great. Let's add a note about why it's OK to change this.

powerjg commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,70 @@
+    }
+)
+# schedule a MAX_INSTS exit event before the simulation begins
+simulator.schedule_max_insts(simpoint.get_simpoint_interval(), True)
```
Can you add a "warmup" period here. I'd like to see how it works to warmup for *x* instructions then run the simpoint for *interval - x* instructions.

Hmm, now that I write that out, maybe it's not quite right. Can you go back to the SimPoint paper and see how they talked about warmup.

studyztp replied:
They mentioned warmup in section 5 and 6. I believe they did the warmup by doing the fast-forwarding: "all
fast forwarding is done so that all of the architecture structures
are completely warmed up when starting simulation (no
cold-start effect)." Later in section 6, the related work section, they mentioned using the entire fast-forward interval for warmup should be the minimum set of fast-forward transactions needed to get to the accurately produce state. They also mentioned checkpoints can either store the architectural state or warm up in a similar manner to the above method.

studyztp replied:
I'm not familiar with warmup but I'll try taking the checkpoint x instructions earlier than the SimPoint and execute the x instructions to perform warmup before taking statistics for the SimPoint. In this case, maybe the number of instructions for warmup should be stored in the SimPoint class too. Please let me know if I misunderstand the warmup process.

powerjg replied:
Yeah, that sounds great. I'm good with adding a warmup instructions parameter to the simpoint object.

powerjg commented on src/cpu/base.cc:
```diff
@@ -727,6 +723,27 @@ BaseCPU::traceFunctionsInternal(Addr pc)
+{
+    std::string cause = "simpoint starting point found";
+    for (size_t i = 0; i < inst_starts.size(); ++i){
+
+        scheduleInstStop(0, inst_starts[i], cause);
+
```
I know I'm being nitpicky here but... can you remove the extra newlines? Also, there should be a space toward the end of the `for` line (`){` -> `) {`)

powerjg commented on src/cpu/base.cc:
```diff
@@ -727,6 +723,27 @@ BaseCPU::traceFunctionsInternal(Addr pc)
+{
+    std::string cause = "a thread reached the max instruction count";
+    for (ThreadID tid = 0; tid < numThreads; ++tid){
+
```
Same as above.

studyztp committed:
stdlib, configs: added warmup support and comments in examples

simpoints-se-checkpoint.py & simpoints-se-restore.py
Changed cache hierarchy to NoCache() when taking checkpoints
Changed memory to simpler memory when taking checkpoints
Added warmup sample
Added comments for cache and memory

base.cc
Fixed style issue

simpoint.py
Added a new parameter to take in warmup interval
Added a function to setup warmup instructions for warmup support
Added a new object in the SimPoint class to store warmup list

Change-Id: I8f66d60b6ed92ac6d0e9841d854803c571a5d10c

aakahlow committed:
mem: Add a flag to disable pkt q size check

This change adds a flag to MemCtrl to allow disabling
response port's queue size sanity check. This is needed
for cases/tests where you might want to drive the memory
system with a much higher bandwidth, for example as in
HBM2.

Change-Id: If8d621339ce8f3ab92cbe2b94039486705c64fc1
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62372
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
stdlib: Fix SwitchableProcessor to allow switch to KVM

An exception was raised if the SwitchableProcessor was setup to
switch to a KVM core from a non-KVM core (i.e., if KVM cores were
present they needed to be the starting core). This was due to a bug in
the Simulator module where the `root.sim_quantum` as not setup for cases
where the SwitchableProcessor was not starting with a KVM core, thus
causing an error when switched to KVM cores.

This has been fixed by modifying the Simulator module to always set
root.sim_quantum. This is acceptable as this is only used in KVM setups.

Change-Id: If57352ba67b7bca81882eae2ef1e9013ef45272f
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62471
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: kokoro <noreply+kokoro@google.com>

studyztp committed:
rebase with the public gem5 develop

studyztp committed:
Delete testing.txt

studyztp committed:
Delete output.txt

hnpl committed:
arch: Bump MaxVecRegLenInBytes to 2^16

Per chapter 2 of RVV 1.0 specs [1], RISCV vector extension supports
vector register size of upto 64kB.

[1] https://github.com/riscv/riscv-v-spec/releases/tag/v1.0

Change-Id: Ib62eb9d59006403f6fe08cfb06cb8a1bc0adbe36
Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62491
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

hnpl committed:
arch-riscv: Remove one of two BasicDecode per atomic insts format

BasicDecode, or decode block templates in general, contains the template
for substituting the return statement returning a StaticInst given a
machine code.

In the case of micro-coding an instruction, this return statement is for
the macro op. Additionally, in gem5 riscv, the spawned micro-ops will be
added in the macro op constructor, which is done in the macro-op
constructor template. Thus, there's no need for having a return statement
for the micro-op.

Currently, there are two return statements in decode-method.cc.inc for
each riscv atomic inst. This change removes one of the two BasicDecode
blocks in atomic inst templates.

This change is expected to a cosmetic change.

Change-Id: Id14bde25d5d3f164b4faafd33bfd5c802a94ca09
Signed-off-by: Hoa Nguyen <hoanguyen@ucdavis.edu>
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62492
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

studyztp committed:
congfigs: added a comment to explain the schedule_max_insts

studyztp committed:
stdlib: moved function from simple_core to base_cpu_core

BaseCPU.py
formatted with python black

abstract_core.py
formatted with python black

simple_core.py
removed simpoint related scheduling functions

base_cpu_core.py
added set_simpoint and set_inst_stop_any_thread

simpoint.py
changed a variable name 'inst_start' to 'start_inst'

This commit was tested with the simpoint config scripts in the
configs/example/gem5_library/checkpoints

Change-Id: I7e30ecc79db3f592e5f9f192c4b936edda0a698a

studyztp committed:
stdlib: formatted the abstract_core.py with python black

Change-Id: I2c1a57d43818f4e352be4143ef4cf8414953b17e

BobbyRBruce commented on src/python/gem5/components/processors/simpoint.py:
```diff
@@ -0,0 +1,171 @@
+        Need to pass in the paths or the lists for the SimPoints and their
+        weights. If the paths are passed in, no actions will be done to the
+        list.
+
```
Perhaps given this usage there should be two `def __init__` functions: one which accepts paths, and another that accepts lists. Is this possible to do? It'd make it much clearer to the user how this is used.

powerjg replied:
I'll push back here. Python doesn't support differentiating functions on input types (it's duck typed!). So, I think this is actually a great way to do it.

You *could* have two function `_init_lists()` and `_init_files()` which the `__init__` calls, but I'm not sure that's any better.

BobbyRBruce replied:
Fair, was just wondering if there was a nice way to do it. If not then go with this.

BobbyRBruce commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +188,42 @@ def unique_exit_event():
+        stage of the core, else, schedule the events during the simulation
+        """
+        if self._board.get_processor().get_num_cores() > 1:
+            warn("SimPoints only work with one core")
```
Should this just be a warning? Maybe it should be a proper exception? A warning won't stop a user doing this, but maybe we should.

powerjg replied:
Good question. It won't *break* anything to have more than 1 core. I think a warning is the right level.

BobbyRBruce commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -66,7 +66,6 @@ def connect_icache(self, port: Port) -> None:
         """
         This function should connect the response port from the instruction
         cache to the right request port on the core.
-
```
I think there's some weird formatting here where these new lines have been removed. Can you re-add them?

BobbyRBruce commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -116,7 +112,38 @@ def connect_interrupt(
         raise NotImplementedError
+
+    @abstractmethod
+    def set_simpoint(
```
Good job working with my latest changes!

Everything here looks good but I'd add a helpful message to the `NotImplementedError`, something like:

```
raise NotImplementedError("This core type does not support simpoints")
```

I'm thinking about the case where we have cores that are not BaseCPUCores. If someone tries to set and use simpoints a vanilla `NotImplementedError` might be confusing.

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,89 @@
+from gem5.components.processors.simpoint import SimPoint
+from pathlib import Path
+from m5.stats import reset, dump
+
```
Maybe just add a bit of documentation here explaining that this should be run after "simpoints-se-checkpoint.py".

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,72 @@
+from gem5.simulate.exit_event import ExitEvent
```
This file is lacking a copyright message.

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,89 @@
+import imp
```
This file is lacking a copyright message.

BobbyRBruce reviewed (changes requested):
Very well done! I have a few questions/comments, but for the most part I agree with most of these changes. It's good work!

BobbyRBruce committed:
misc: Add the legacy gem5 .git/hooks to the pre-commit

The gem5 commit msg checker must be run during the 'commit-msg'
stage ergo this is explicitly set. The other hooks are only applicable
to the "commit" stage, the `default_stage` for the hooks.

To install all the hooks, you need to run the following:

```
pre-commit install -t pre-commit -t commit-msg
```

This ensures both the 'commit-msg' and 'pre-commit' hooks are installed.
If you run just `pre-commit install`, only the pre-commit hooks are
installed.

Change-Id: I4a0dcc7159ed5048baa120adf80bbf65f63c11dd
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62552
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

KyleRoarty committed:
configs: Add CLA for max_cu_tokens to apu_se.py

Adds a command-line argument for the Compute Unit's maximum
number of coalescer tokens

Change-Id: Ie00c3ac7e28ad162b801643bff345b4cf434a878
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62551
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Matthew Poremba <matthew.poremba@amd.com>
Reviewed-by: Matt Sinclair <mattdsinclair@gmail.com>
Reviewed-by: Matthew Poremba <matthew.poremba@amd.com>
Maintainer: Matt Sinclair <mattdsinclair@gmail.com>

BobbyRBruce committed:
scons: Update automatic hook install for pre-commit

This replaces the old hooks with the pre-commit check. If Python
pre-commit has not been installed an error is thrown asking the user to
install the requirements via pip.

Change-Id: I2d42f42624e10d38d0da39b473f0363db128ce1c
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62553
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

studyztp committed:
stdlib: moved simpoint.py from processor dir to utils

simpoints-se-checkpoint.py & simpoint-se-restore.py:
added explaination and usage on the top of the scripts
changed SimPoint module directory

SConscript:
changed SimPoint module directory

se_binary_workload.py:
changed SimPoint module directory

abstract_core.py:
fixed the format

simpoint.py:
moved dirctory from processor to utils

This commit is tested on the simpoint-se-checkpoint.py and
the simpoint-se-restore.py

Change-Id: Ib01f097660855a657bb1f504297ae37495813c7d

studyztp committed:
stdlib: fixed the error message of MAX_INSTS schedule function

Change-Id: Iffe00e8beea59f4bb35392c12d3173e8c5f71d6b

jiegec committed:
configs: Add option to override cpu vendor string

Glibc requires x86-64-v2 ISA level on newer Linux distributions (e.g.
Debian Bookworm), and running applications in GEM5 will fail with "CPU
ISA level is lower than required" error. It is due to glibc not
detecting CPU features when the vendor string is unknown yet requiring
them to run. For glibc to detect correct CPU features, this commit adds
a command line option to allow user to override x86 cpu vendor string to
well-known ones, e.g. GenuineIntel. It allows glibc to detect more cpu
features and fixes the issue.

Jira Issue: https://gem5.atlassian.net/browse/GEM5-1117

Change-Id: I22907e7b983e9aa6122543042af207e35b09badb
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62555
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
stdlib: Fix SimpleCore no ISA set use-case

The error was highlighted via this Nightly test failure:
https://jenkins.gem5.org/job/nightly/324

The bug was triggered when creating a SimpleCore without passing an ISA.
This is allowed but the SimpleCore should have ran `get_runtime_isa` to
determine which ISA is to be used in the simulation. This was missing.
This patch fixes this.

Change-Id: I4b2718233819783f779462d24b694306e9e76e30
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62571
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
misc: Update CONTRIBUTING.md for `pre-commit`

This patch does two things:

1. Ensures that both the 'pre-commit' and 'commit-msg' hooks are
   installed. The `pre-commit install` by itself will only install the
   'pre-commit' hooks. This expanded instlal command will also install
   the 'commit-msg' hook.
2. How to run pre-commit to automatically format your code.

Change-Id: I0561f2918568bb9191e4ec457c297fcd264248c0
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62573
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

BobbyRBruce committed:
misc: Add gerrit commit message hook to pre-commit

This pre-commit hook adds the Change-ID etc. to the commit message.

Change-Id: I43b239468ab60ce1a8aaabb53dd6905819a2ce05
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62575
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>
Tested-by: kokoro <noreply+kokoro@google.com>

jiegec committed:
arch-arm: Switch from getVec* to getReg* accessor.

Commit 7f61db0f missed one change in armv8_cpu.cc and led to
compilation error. The fix is done in the same way as
remote_gdb.cc in the previous commit.

This commit also fixes undeclared error by adding includes.

Jira issue: https://gem5.atlassian.net/browse/GEM5-1271

Change-Id: I48ce9b430d1d1988d5595ccfcbd7175a6cd030b7
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62591
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

jiegec committed:
misc: Fix dynamic decision of TranslatingPortProxy.

In commit 83b14e56, getVirtProxy is replaced by inline ternary operators
to decide between FS or SE version. However, dynamic dispatch will not
work in this scenario and the virtual function of SETranslatingPortProxy
will not be called. It may lead to failure in m5op read_file in SE mode.

Change-Id: I9b5f757096cfdbd6fb8bc14b1b0e02245703a0ac
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62611
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
misc: Exclude test ref directories from pre-commit

The files in "tests/*./ref" directories are used in tests to check the
output of a test is valid. As such these should not be automatically
formatted by the pre-commit.

Change-Id: I82be1a91132a0b6c66c8bbb8f6d7dcc6e72abe77
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62631
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
python: Add 'requirements.txt'

Change-Id: Ic7ef47bbef8879d6235ef669ff39a17f602f5cfd
Issue-on: https://gem5.atlassian.net/browse/GEM5-1268
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62632
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

BobbyRBruce committed:
misc: Run pre-commit run on all files in repo

The following command was run:

```
pre-commit run --all-files
```

This ensures all the files in the repository are formatted to pass our
checks.

Change-Id: Ia2fe3529a50ad925d1076a612d60a4280adc40de
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62572
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

powerjg committed:
stdlib: Improve core names in switchable processor

Currently, when using the switchable processor the first N cores are the
starting cores and the next N cores (e.g., board.processor.core<N+1>)
are the switched in cores. This is confusing when looking at the stats.

This change makes it so that the names of the different processor lists
used in the dictionary when constructing the switchable processor are
used in for the member names as well. This will allow users to have
names like board.processor.ff_cores and board.processor.detailed_cores.

A bit of refactoring of the base processor was required for this.

Change-Id: I244ee5f6080599acb60a777da979da048cf7463e
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62652
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

KyleRoarty committed:
configs: Add default to max_cu_tokens CLA

Fixes nightly tests

Change-Id: Ibad2e8a52afb62b0605ffa7ca958e378df799dae
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62653
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Matt Sinclair <mattdsinclair@gmail.com>
Maintainer: Matt Sinclair <mattdsinclair@gmail.com>

youngin-abel-kim committed:
stdlib: Minor typo fixed

In hbm.py, it says "Interfaces for LPDDR5 memory devices".

I think LPDDR5 should be replaced with HBM.

Change-Id: I87e0beafa79e6e3d9176edaf69b34a38230e9271
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62654
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Jiajie Chen <c@jia.je>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
stdlib: Remove setting of 'kvm_vm' from SwitchableProcessor

This stops a `kvm_mv already has parent` warning from happening when
using a SwitchableProcessor.

Change-Id: I495a040e03c33228ceafc99a94b0d0957f4ff6a2
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62657
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

powerjg committed:
python: Fix up arrow in interactive shell

Simply by importing the readline module, the up arrow will work in
gem5's interactive shell now.

Change-Id: I41d87adc34253abf5a00ac484da377f9f065a27a
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62671
Tested-by: kokoro <noreply+kokoro@google.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

studyztp committed:
stdlib: added back some changes in simple_core.py

The rebase I've done a week ago didn't merge changes in the
simple_core.py.

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -82,5 +89,12 @@ def set_se_binary_workload(
 
+        if simpoint is not None:
+          if self.get_processor().get_num_cores() > 1:
+            warn("SimPoints only works with one core")
```
Needs two more spaces (run black?)

powerjg commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -120,3 +120,35 @@ def get_mmu(self) -> BaseMMU:
+
+    @abstractmethod
+    def set_simpoint(
+        self, inst_starts: List[int] = [], init: bool = True
```
Why do these parameters have default values? Are neither of them required? `inst_starts` seems required at a minimum.

powerjg commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -120,3 +120,35 @@ def get_mmu(self) -> BaseMMU:
+
+    @abstractmethod
+    def set_inst_stop_any_thread(
+        self, inst: int = 0, init: bool = True
```
Why do these parameters have default values? Are neither of them required? `inst` seems required at a minimum as it doesn't make sense to stop at instruction 0, right?

powerjg commented on src/python/gem5/components/processors/base_cpu_core.py:
```diff
@@ -151,3 +151,21 @@ def connect_interrupt(
+
+    @overrides(AbstractCore)
+    def set_simpoint(
+        self, inst_starts: List[int] = [], init: bool = True
```
Ditto about default values.

powerjg commented on src/python/gem5/components/processors/base_cpu_core.py:
```diff
@@ -151,3 +151,21 @@ def connect_interrupt(
+
+    @overrides(AbstractCore)
+    def set_inst_stop_any_thread(
+        self, inst: int = 0, init: bool = True
```
Same question on default values.

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -103,3 +104,33 @@ def default_workend_generator():
+        yield False
+
+def simpoint_dump_reset_generator():
+    while True:
+        """
+            A generator for SimPoints. It will reset the simulation statistics
+            and then dump simulation statistics.
+        """
```
```suggestion
        """
            A generator for SimPoints. It will reset the simulation statistics
            and then dump simulation statistics.
        """
    while True:
```

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -103,3 +104,33 @@ def default_workend_generator():
+        yield False
+
+def simpoint_save_checkpoint_generator(checkpoint_dir: Path):
+    while True:
+        """
+            A generator for SimPoints. It will take a checkpoint with the input
+            path and the current simulation Ticks.
+        """
```
```suggestion
        """
            A generator for SimPoints. It will take a checkpoint with the input
            path and the current simulation Ticks.
        """
    while True:
```

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -103,3 +104,33 @@ def default_workend_generator():
+    while True:
+        yield False
+
+def simpoint_dump_reset_generator():
```
Does this have to be "simpoint" specific, or is this just a dump/reset generator?

powerjg commented on src/python/gem5/simulate/exit_event_generators.py:
```diff
@@ -103,3 +104,33 @@ def default_workend_generator():
+        m5.stats.reset()
+        yield False
+
+def simpoint_save_checkpoint_generator(checkpoint_dir: Path):
```
Ditto, is this simpoint-specific?

powerjg commented on src/python/gem5/utils/simpoint.py:
```diff
@@ -0,0 +1,171 @@
+        simpoint_list: List[int] = None,
+        weight_list: List[int] = None,
+        warmup_interval: int = 0
+        # bbv_file_path: Path = None, will take in bbv and generate simpoints
+        # with it in the future
```
Do we really need to commit to this? I suggest removing this comment

powerjg commented on src/python/gem5/utils/simpoint.py:
```diff
@@ -0,0 +1,171 @@
+        if warmup_interval != 0:
+            self._warmup_list = self.set_warmup_intervals(warmup_interval)
+        else:
+            self._warmup_list = [
+                0 for _ in range(len(self._simpoint_start_insts))
+            ]
```
```suggestion
            self._warmup_list = [0] * len(self._simpoint_start_insts)
```

powerjg reviewed (commented):
I think the only thing left to do is to split this into specific commits. Other than the small comments that remain, we're almost ready to submit to gerrit!

powerjg committed:
configs: Fix stat names after switchable changes

b6e0e72d9 changed the names of the switchable processor cores. This
change updates the stats after the nightlies failed.

Change-Id: If349ff07dea08ad3999e02ee95da389bab903b3e
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5/+/62791
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: kokoro <noreply+kokoro@google.com>

studyztp committed:
stdlib: remove some default values and fix the fucntion names

simpoints-se-checkpoint.py:
Changed the generator name

se_binary_workload.py:
passing init value to the schedule_simpoint function

abstract_core.py:
added ":bool" to init

base_cpu_core.py:
removed default values for set_simpoint and set_inst_stop_any_thread

exit_event_generators.py:
change dump/reset and checkpoint generator from simpoint specific
to not simpoint specific

simulator.py:
removed default value for insts and start_insts for simpoint and
max-inst scheduling functions

simpoint.py:
changed the way to init warmup list
applied python black format

Change-Id: Ia83c2d6b68dce6ea034b53ab73754e4dfa2c48a8

studyztp commented:
@powerjg How about splitting them into an exit_event_generator commit, a simpoints module commit, a base cpu simpoints related scheduling function commit, and a stdlib simpoints support commit(the simulator, exit event, and se_set_workload)?
